File: src/executor/filter.ts
import {Iter, RowMarker} from "./iter";

export class Filter implements Iter
{
    private source: Iter;
    private property: string;
    private value: any;
    private columns: string[];

    constructor(source: Iter, columns: string[], property: string, value: any) {
        this.source = source;
        this.property = property;
        this.value = value;
        this.columns = columns;
    }

    async next(): Promise<any[] | RowMarker> {
        const row = await this.source.next();
        switch (row) {
            case "end of file" as RowMarker: return row;
            case "empty row" as RowMarker: return row;
            default: return (row as any[])[this.columns.indexOf(this.property)] === this.value ? row : this.next();
        }
    }

}

File: rust-toydb/src/filter.rs
use crate::producer::{Itrator, Row};

pub(crate) struct Filter<'a> {
    source: &'a mut dyn Itrator,
}

impl<'a> Filter<'a> {
    pub(crate) fn new(source: &'a mut dyn Itrator) -> Filter {
        Filter {
            source
        }
    }
}

impl<'a> Itrator for Filter<'a> {
    fn next(&mut self) -> Option<Row> {
        self.source.next()
    }
}